File: Examples/matlab/runMTF.m
function runMTF(varargin)

close all;

% script parameters
use_rgb_input = 1;
init_from_gt = 1;
show_fps = 0;
show_window = 0;

% MTF parameters
config_dir = '../../Config';
db_root_path = '../../../Datasets';
pipeline = 'c';
img_source = 'j';
actor_id = 1;
seq_id = 16;
seq_fmt = 'jpg';
init_frame_id = 1;

param_str = sprintf('db_root_path %s', db_root_path);
param_str = sprintf('%s pipeline %s', param_str, pipeline);
param_str = sprintf('%s img_source %s', param_str, img_source);
param_str = sprintf('%s actor_id %d', param_str, actor_id);
param_str = sprintf('%s seq_id %d', param_str, seq_id);
param_str = sprintf('%s seq_fmt %s', param_str, seq_fmt);
param_str = sprintf('%s init_frame_id %d', param_str, init_frame_id);
if mod(nargin, 2) ~= 0
    error('Optional arguments must be specified in pairs');
end
% parse optional arguments
arg_id = 1;
while arg_id <= nargin    
    arg_name = varargin{arg_id};
    arg_val = varargin{arg_id + 1};
    if strcmp(arg_name, 'use_mtf_pipeline')
        use_mtf_pipeline = arg_val;
    elseif strcmp(arg_name, 'init_from_gt')
        init_from_gt = arg_val;
    elseif strcmp(arg_name, 'show_fps')
        show_fps = arg_val;
    elseif strcmp(arg_name, 'show_window')
        show_window = arg_val;
    elseif strcmp(arg_name, 'config_dir')
        config_dir = arg_val;
    else
        param_str = sprintf('%s %s %s', param_str,...
            string(arg_name), string(arg_val));
    end
    arg_id = arg_id + 2;
end   
param_str = sprintf('config_dir %s %s', config_dir, param_str);

[sequences, actors] = datasets;
actor = actors{actor_id+1};
seq_name = sequences{actor_id + 1}{seq_id + 1};
seq_path = sprintf('%s/%s/%s', db_root_path, actor, seq_name);

[input_id, n_frames] = mexMTF('init', param_str);
if input_id == 0
    error('MTF input pipeline creation was unsuccessful');
else
    fprintf('MTF input pipeline created successfully');
    if n_frames > 0
        fprintf(' with %d frames', n_frames);
    end
    fprintf('\n');
end

tracker_id = mexMTF('create_tracker', param_str);
if ~tracker_id
    error('Tracker creation was unsuccessful');
else
    fprintf('Tracker created successfully\n');
end

if ~success
    error('Tracker initialization was unsuccessful');
else
    fprintf('Tracker initialized successfully\n');
end
% pause
avg_error = 0;
frame_id = init_frame_id + 1;
if show_window
    h = figure; 
end
while 1
    [success, curr_img] = mexMTF('get_frame', input_id);
    if ~success
        error('MTF input pipeline update was unsuccessful for frame %d', frame_id + 1);
    end   
    if ~use_rgb_input
        curr_img = rgb2gray(curr_img);
    end
    start_t = tic;
    [success, curr_corners] = mexMTF('get_region', tracker_id, uint8(curr_img));
    time_passed = toc(start_t);
    if ~success
        error('Tracker update was unsuccessful');
    end
	if show_fps
		fps = 1.0 /double(time_passed);
		fprintf('time_passed: %f\t fps: %f\n',time_passed, fps);
    end
	if show_window
        if ~ ishghandle(h)
            break;
        end
		imshow(curr_img);
		hold on;
		plot([curr_corners(1, :),curr_corners(1, 1)],...
            [curr_corners(2, :), curr_corners(2, 1)],...
            'Color','r', 'LineWidth',2);
        if init_from_gt
            plot(gt_data(frame_id, [1, 3, 5, 7, 1]),...
                gt_data(frame_id, [2, 4, 6, 8, 2]),...
                'Color','g', 'LineWidth',2);
        end
		hold off;
		pause(0.001);
    else
		if mod(frame_id, 50)==0
			fprintf('frame_id: %d\t', frame_id);
            if init_from_gt
                gt_corners(1, :) = gt_data(frame_id, [1, 3, 5, 7]);
                gt_corners(2, :) = gt_data(frame_id, [2, 4, 6, 8]);
                curr_error = norm((gt_corners - curr_corners), 'fro')';
                avg_error = avg_error + (curr_error - avg_error)/(frame_id - init_frame_id);
                fprintf('\t curr_error: %f\t avg_error: %f',...
                    frame_id, curr_error, avg_error);
            end 
            fprintf('\n');
		end
	end
    frame_id = frame_id + 1;
    if n_frames > 0 && frame_id >= n_frames
        break;
    end
end
mexMTF('quit');

File: Examples/python/runMTF.py
import os
import sys
import cv2
import numpy as np
import math
import time
import pyMTF
from utilities import readGroundTruth, writeCorners, drawRegion
from datasets import sequences, actors

if __name__ == '__main__':

    config_root_dir = '../../Config'
    db_root_dir = '../../../../../Datasets'
    actor_id = 0
    seq_id = 0
    seq_fmt = 'jpg'
    use_rgb_input = 0

    write_stats_to_file = 0
    show_tracking_output = 1

    arg_id = 1
    if len(sys.argv) > arg_id:
        seq_id = int(sys.argv[arg_id])
        arg_id += 1
    if len(sys.argv) > arg_id:
        write_stats_to_file = int(sys.argv[arg_id])
        arg_id += 1
    if len(sys.argv) > arg_id:
        show_tracking_output = int(sys.argv[arg_id])
        arg_id += 1

    if seq_id >= len(sequences):
        print('Invalid dataset_id: ', seq_id)
        sys.exit()

    actor = actors[actor_id]
    seq_name = sequences[actor][seq_id]
    print('seq_id: ', seq_id)
    print('seq_name: ', seq_name)

    src_fname = '{:s}/{:s}/{:s}/frame%05d.{:s}'.format(db_root_dir, actor, seq_name, seq_fmt)
    ground_truth_fname = '{:s}/{:s}/{:s}.txt'.format(db_root_dir, actor, seq_name)
    result_fname = seq_name + '_res.txt'

    cap = cv2.VideoCapture()
    if not cap.open(src_fname):
        print('The video file ', src_fname, ' could not be opened')
        sys.exit()

    # thickness of the bounding box lines drawn on the image
    thickness = 2
    # ground truth location drawn in green
    ground_truth_color = 'green'
    # tracker location drawn in red
    result_color = 'red'

    # read the ground truth
    ground_truth = readGroundTruth(ground_truth_fname)
    no_of_frames = ground_truth.shape[0]

    print('no_of_frames: ', no_of_frames)

    ret, init_img = cap.read()
    if not ret:
        print('Initial frame could not be read')
        sys.exit(0)

    # extract the true corners in the first frame and place them into a 2x4 array
    init_corners = [ground_truth[0, 0:2].tolist(),
                    ground_truth[0, 2:4].tolist(),
                    ground_truth[0, 4:6].tolist(),
                    ground_truth[0, 6:8].tolist()]
    init_corners = np.array(init_corners).T
    # print('init_corners:\n {}'.format(init_corners))

    # initialize tracker with the first frame and the initial corners
    if not use_rgb_input:
        init_img = cv2.cvtColor(init_img, cv2.COLOR_RGB2GRAY)
        tracker_id = pyMTF.create(init_img.astype(np.uint8), init_corners.astype(np.float64), config_root_dir)
    if not tracker_id:
        print('Tracker creation/initialization was unsuccessful')
        sys.exit()

    if show_tracking_output:
        # window for displaying the tracking result
        window_name = 'Tracking Result'
        cv2.namedWindow(window_name)

    # lists for accumulating the tracking error and fps for all the frames
    tracking_errors = []
    tracking_fps = []

    tracker_corners = np.zeros((2, 4), dtype=np.float64)

    for frame_id in range(1, no_of_frames):
        ret, src_img = cap.read()
        if not ret:
            print("Frame ", frame_id, " could not be read")
            break
        actual_corners = [ground_truth[frame_id, 0:2].tolist(),
                          ground_truth[frame_id, 2:4].tolist(),
                          ground_truth[frame_id, 4:6].tolist(),
                          ground_truth[frame_id, 6:8].tolist()]
        actual_corners = np.array(actual_corners).T

        if not use_rgb_input:
            src_img_disp = src_img.copy()
            src_img = cv2.cvtColor(src_img, cv2.COLOR_RGB2GRAY)
        else:
            src_img_disp = src_img

        start_time = time.clock()

        # update the tracker with the current frame
        success = pyMTF.getRegion(src_img.astype(np.uint8), tracker_corners, tracker_id)

        if not success:
            print('Tracker update was unsuccessful')
            sys.exit()

        # print('tracker_corners before:\n {}'.format(tracker_corners))
        # success = pyMTF.setRegion(tracker_corners, tracker_id)
        # print('tracker_corners after:\n {}\n\n'.format(tracker_corners))

        end_time = time.clock()

        # compute the tracking fps
        current_fps = 1.0 / (end_time - start_time)
        tracking_fps.append(current_fps)

        # compute the tracking error
        current_error = math.sqrt(np.sum(np.square(actual_corners - tracker_corners)) / 4)
        tracking_errors.append(current_error)

        if show_tracking_output:
            # draw the ground truth location
            drawRegion(src_img_disp, actual_corners, ground_truth_color, thickness)
            # draw the tracker location
            drawRegion(src_img_disp, tracker_corners, result_color, thickness)
            # write statistics (error and fps) to the image
            cv2.putText(src_img_disp, "{:5.2f} {:5.2f}".format(current_fps, current_error), (5, 15),
                        cv2.FONT_HERSHEY_COMPLEX_SMALL, 1, (255, 255, 255))
            # display the image
            cv2.imshow(window_name, src_img_disp)

            if cv2.waitKey(1) == 27:
                break

        pyMTF.remove(tracker_id)

    mean_error = np.mean(tracking_errors)
    mean_fps = np.mean(tracking_fps)

    print('mean_error: ', mean_error)
    print('mean_fps: ', mean_fps)
